app/core/DeeplinkManager/ParseManager/utils/verifySignature.ts

```diff
@@ -36,14 +36,40 @@ function canonicalize(url: URL): string {
 
   params.delete('sig');
 
-  params.sort();
-
-  const queryString = params.toString();
-
-  const fullUrl =
-    url.origin + url.pathname + (queryString ? `?${queryString}` : '');
+  let queryString = '';
+
+  if (params.has('sig_params')) {
+    const sigParamsStr = params.get('sig_params');
+    const allowedParams = sigParamsStr?.split(',') ?? [];
+    const signedParams = new URLSearchParams();
+    const existingParams: string[] = [];
+    const seenParams = new Set<string>(); // Track which params we've already processed
+
+    for (const param of allowedParams) {
+      if (params.has(param) && !seenParams.has(param)) {
+        const value = params.get(param);
+        if (value !== null) {
+          signedParams.set(param, value);
+          existingParams.push(param);
+          seenParams.add(param); // Mark this param as processed
+        }
+      }
+    }
+    // CRITICAL: Include normalized sig_params in the signature to prevent tampering
+    // Only include params that actually exist in the URL (deduplicated and sorted)
+    existingParams.sort(); // Sort parameter names alphabetically for consistent canonicalization
+    const rejoinedSigParams = existingParams.join(',');
+    signedParams.set('sig_params', rejoinedSigParams);
+
+    signedParams.sort();
+    queryString = signedParams.toString();
```

Comment:
S: url params can have whitespaces encoded in two ways: `%20` vs `+`. (and other unicode chars but they are often encoded the same way with %XX)
If we don't normalise whitespaces encoding here, url returned as `canonicalUrl` can vary even if we have the same spaces. So we should [normalise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize) the spaces by forcing it.

```suggestion
    queryString = signedParams.toString().normalize('NFC');
```